nemo/collections/asr/modules/rnn_encoder.py

```diff
@@ -0,0 +1,187 @@
+        dropout=0.1,
+    ):
+        super().__init__()
+
+        self.d_model = d_model
+        self._feat_in = feat_in
+
+        if subsampling_conv_channels == -1:
+            subsampling_conv_channels = proj_size
+        if subsampling and subsampling_factor > 1:
+            if subsampling == 'stacking':
+                self.pre_encode = StackingSubsampling(
+                    subsampling_factor=subsampling_factor, feat_in=feat_in, feat_out=proj_size
+                )
+            else:
+                self.pre_encode = ConvSubsampling(
+                    subsampling=subsampling,
+                    subsampling_factor=subsampling_factor,
+                    feat_in=feat_in,
+                    feat_out=proj_size,
+                    conv_channels=subsampling_conv_channels,
+                    activation=nn.ReLU(),
+                )
+        else:
+            self.pre_encode = nn.Linear(feat_in, proj_size)
+
+        self._feat_out = proj_size
+
+        self.layers = nn.ModuleList()
+
+        SUPPORTED_RNN = {"lstm": nn.LSTM, "gru": nn.GRU, "rnn": nn.RNN}
+        if rnn_type not in SUPPORTED_RNN:
+            raise ValueError(f"rnn_type can be one from the following:{SUPPORTED_RNN.keys()}")
+        else:
+            rnn_module = SUPPORTED_RNN[rnn_type]
+
+        for i in range(n_layers):
+            rnn_proj_size = proj_size // 2 if bidirectional else proj_size
+            if rnn_type == "lstm":
+                layer = rnn_module(
+                    input_size=self._feat_out,
+                    hidden_size=d_model,
+                    num_layers=1,
+                    batch_first=True,
+                    bidirectional=bidirectional,
+                    proj_size=rnn_proj_size,
+                )
+            self.layers.append(layer)
+            self.layers.append(nn.LayerNorm(proj_size))
+            self.layers.append(nn.Dropout(p=dropout))
+            self._feat_out = proj_size
+
+        if proj_size > 0 and self._feat_out != proj_size:
+            self.out_proj = nn.Linear(self._feat_out, proj_size)
+            self._feat_out = proj_size
+        else:
+            self.out_proj = None
+
+    @typecheck()
+    def forward(self, audio_signal, length=None):
+        max_audio_length: int = audio_signal.size(-1)
+
+        if length is None:
+            length = audio_signal.new_full(
+                audio_signal.size(0), max_audio_length, dtype=torch.int32, device=self.seq_range.device
+            )
+
+        audio_signal = torch.transpose(audio_signal, 1, 2)
+
+        if isinstance(self.pre_encode, ConvSubsampling) or isinstance(self.pre_encode, StackingSubsampling):
+            audio_signal, length = self.pre_encode(audio_signal, length)
+        else:
+            audio_signal = self.pre_encode(audio_signal)
+
+        for lth, layer in enumerate(self.layers):
+            if isinstance(layer, torch.nn.Dropout) or isinstance(layer, torch.nn.LayerNorm):
```

Comment:
For cleanliness, how about doing outputs = layer(audio_signal) and then check if type(outputs) in (list, tuple), then unpack the 0th element for LSTM/GRU/RNN etc?

Reply:
Good idea! changed it.